tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/AWS/SQS/AwsSqsHandlerCommon.cs

```diff
@@ -0,0 +1,170 @@
+
+        var scope = AwsSqsCommon.CreateScope(Tracer.Instance, sendType.OperationName, out var tags, spanKind: SpanKinds.Producer);
+
+        var queueName = AwsSqsCommon.GetQueueName(requestProxy.QueueUrl);
+        if (tags is not null && requestProxy.QueueUrl is not null)
+        {
+            tags.QueueUrl = requestProxy.QueueUrl;
+            tags.QueueName = queueName;
+        }
+
+        if (scope?.Span.Context != null && !string.IsNullOrEmpty(queueName))
+        {
+            var dataStreamsManager = Tracer.Instance.TracerManager.DataStreamsManager;
+
+            if (sendType == SendType.SingleMessage)
+            {
+                InjectForSingleMessage(dataStreamsManager, request, scope, queueName);
+            }
+            else if (sendType == SendType.Batch)
+            {
+                InjectForBatch(dataStreamsManager, request, scope, queueName);
+            }
+        }
+
+        return new CallTargetState(scope);
+    }
+
+    private static void InjectForSingleMessage<TSendMessageRequest>(DataStreamsManager dataStreamsManager, TSendMessageRequest request, Scope scope, string queueName)
+    {
+        var requestProxy = request.DuckCast<ISendMessageRequest>();
+        if (dataStreamsManager != null && dataStreamsManager.IsEnabled)
+        {
+            var edgeTags = new[] { "direction:out", $"topic:{queueName}", "type:sqs" };
+            scope.Span.SetDataStreamsCheckpoint(dataStreamsManager, CheckpointKind.Produce, edgeTags, payloadSizeBytes: 0, timeInQueueMs: 0);
+        }
+
+        ContextPropagation.InjectHeadersIntoMessage<TSendMessageRequest>(requestProxy, scope.Span.Context, dataStreamsManager);
+    }
+
+    private static void InjectForBatch<TSendMessageBatchRequest>(DataStreamsManager dataStreamsManager, TSendMessageBatchRequest request, Scope scope, string queueName)
+    {
+        var requestProxy = request.DuckCast<ISendMessageBatchRequest>();
+        if (requestProxy == null)
+        {
+            return;
+        }
+
+        var edgeTags = new[] { "direction:out", $"topic:{queueName}", "type:sqs" };
+        foreach (var e in requestProxy.Entries)
+        {
+            var entry = e.DuckCast<IContainsMessageAttributes>();
+            // this has no effect is DSM is disabled
+            scope.Span.SetDataStreamsCheckpoint(dataStreamsManager, CheckpointKind.Produce, edgeTags, payloadSizeBytes: 0, timeInQueueMs: 0);
+            // this needs to be done for context propagation even when DSM is disabled
+            // (when DSM is enabled, it injects the pathway context on top of the trace context)
+            ContextPropagation.InjectHeadersIntoMessage<TSendMessageBatchRequest>(entry, scope.Span.Context, dataStreamsManager);
+        }
+    }
+
+    internal static TResponse AfterSend<TResponse>(TResponse response, Exception exception, in CallTargetState state)
+    {
+        state.Scope.DisposeWithException(exception);
+        return response;
+    }
```

Comment:
nit: Probably doesn't need to be a separate method tbh, I would just call `state.Scope.DisposeWithException(exception);` in the integration directly, but I get why you'd like this for symmetry 🙂

Reply:
Yes, I'd prefer having everything here for consistency :3